specification/dartLangSpec.tex

```diff
@@ -14793,71 +14794,151 @@ \subsubsection{For-in}
 which is not a subtype of \code{Iterable<dynamic>}.
 }
 
+\LMHash{}%
+The scope introduced by \code{\{\,$S$\,\}} around $S$
+is known as the
+\IndexCustom{scope of the for-in statement $s$}{%
+  scope!for-in statement}.
+
 
 \subsubsection{Asynchronous For-in}
 \LMLabel{asynchronousFor-in}
 
 \LMHash{}%
 A for-in statement may be asynchronous.
 The asynchronous form is designed to iterate over streams.
-An asynchronous for loop is distinguished by the keyword \AWAIT{} immediately preceding the keyword \FOR.
+An asynchronous for loop is distinguished by
+the keyword \AWAIT{} immediately preceding the keyword \FOR.
+
+\LMHash{}%
+It is a compile-time error if an asynchronous for-in statement appears
+inside a synchronous function (\ref{functions}).
+It is a compile-time error if a traditional for loop
+(\ref{forLoop})
+is prefixed by the \AWAIT{} keyword.
+
+\rationale{%
+An asynchronous loop would make no sense within a synchronous function,
+for the same reasons that an await expression makes no sense in
+a synchronous function.%
+}
+
+\LMHash{}%
+Let $D$ be derived from \syntax{<finalConstVarOrType>?}\ and
+let $s$ be an asynchronous for-in statement of the form
+`\code{\AWAIT\,\,\FOR\,\,($D$\,\,\id\,\,\IN\,\,$e$)\,\,$s_b$}'.
+Let $T_e$ be the static type of $e$.
+A compile-time error occurs unless $T_e$ may be assigned to
+\code{Stream<\DYNAMIC>}.
+%% TODO(eernst): Come nnbd, `a top type` --> `\DYNAMIC{}`.
+If $T_e$ is a top type then let $T_{\metavar{elm}}$ be \DYNAMIC.
+Otherwise there exists a type $U$
+such that $T_e$ implements \code{Stream<$U$>},
+in which case $T_{\metavar{elm}}$ is $U$.
+
+\LMHash{}%
+If $D$ is empty:
+Assume that $v$ is a fresh variable with declared type $T_{\metavar{elm}}$.
+%% Several cases here: \id{} could denote a local variable, an instance or
+%% static setter, nothing, or a lot of things causing an error.
+A compile-time error occurs if replacing $s$ by
+the statement `\code{\id\,\,=\,\,$v$;}'
+would give rise to a compile-time error.
```

Comment:
Too vague. 

For one thing, you at least have to say *which kind of variable* `v` is. Local would be most reasonable. Then you have to say that it must be definitely assigned. 
Then we have to ensure that removing/replacing the loop statement doesn't introduce an error. For example, if the variable `\id` is defined as `late final int id;` and the loop is followed by an `id = 42;` then replacing the loop which *potentially* initializes `id` with an assignment which *definitely* assigns `id` will make the later assignment a compile-time error. That code is possible because the loop might break past the later assignment in the first iteration. Consider the function:
```dart
T first<T>(Stream<T> stream, T defaultValue) {
  late final T result;
  search: {
    await for (result in stream) break search;
    result = defaultValue;
 }
 return result;
}
```
This is "perfectly good" (perhaps a little over-engineered) code. The clause above would disallow it.

So, no removing or replacing code unless it's to something definitely semantically equivalent.
Maybe:

> ```latex
> Replace $S$ with:
> \code{\AWAIT{} \FOR{} ($T_{\metavar{elm}}$ $v$ \IN{} $e$) \{ 
>     \id = $v$; \{ $s_b$; \} 
> \}}
> where $v$ is a fresh variable. It is a compile time error if this would introduce a new compile-time error.
> ```

?
This rewrite should work whether `D` is empty or not.

Reply:
We use a `fresh variable` in a similar way in 36 locations, and we define that concept with the added information that it is assumed to be local.

We do not currently include null safety specific information (such as 'this variable is definitely assigned'). I added a TODO comment on the introduction of the concept, such that we will get that part as part of the concept as well.

About the statement replacement: We can fiddle with the rewriting template, but it doesn't seem useful to me to come up with a complex piece of synthetic code if a simpler piece of code would do.

So we might instead be able to say explicitly that the question is whether the  assignment statement itself is an error, and it is just irrelevant whether the actual replacement would create a program that has some errors else where. I tried to do that.

I don't see the need to require that this desugaring operation yields a semantically equivalent program, because the purpose of the desugaring is to give a compact and readable way to determine whether the original code has a compile-time error, there is no intention of running this code.